HTML API: Ensure that full processor can seek to earlier bookmarks

sirreal opened the pull request:
[#62290](https://core.trac.wordpress.org/ticket/62290) describes an issue where the full processor cannot seek to earlier bookmarks.

When the processor seeks to an earlier place, it returns the the beginning of the document and proceeds forward until it reaches the appropriate location. This requires resetting internal state so that the processor can correctly proceed from the beginning of the document.

The seeking reset logic was not adapted to account for the full processor. This change updates the seek logic to account for the full and fragment parsers as well as other state that has been introduced in the interim and should be reset.

Trac ticket: https://core.trac.wordpress.org/ticket/62290

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

sirreal committed:
Add failing test case

sirreal committed:
Work on fix

sirreal committed:
wip

sirreal committed:
try fixes

sirreal committed:
Fix the issue

sirreal committed:
Restore original bookmark matching

sirreal committed:
Lint: remove empty line

sirreal committed:
Add commentary around state resetting in seek

sirreal committed:
Use do-while loop to iterate after first check

sirreal committed:
Fix fragment parser bug that should reset breadcrumbs before seeking

sirreal committed:
Assert setting bookmark two

This is consisten with the assertion on bookmark one

sirreal committed:
Remove condition that may not be necessary

This seems related to virutal tokens and it likely covered by the virtual token condition

sirreal committed:
Set parsing namespace correctly when seeking

sirreal commented:
There are a lot of parallels with https://github.com/WordPress/wordpress-develop/pull/7348 where the state should be reset appropriately according to the context node.

sirreal committed:
Fix fragment state reset for different context nodes

sirreal committed:
Add and improve comments

sirreal commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -5343,32 +5342,65 @@ public function seek( $bookmark_name ): bool {
+				$this->change_parsing_namespace(
+					$this->context_node->integration_node_type
+						? 'html'
+						: $this->context_node->namespace
+				);
+
+				if ( 'TEMPLATE' === $this->context_node->node_name ) {
+					$this->state->stack_of_template_insertion_modes[] = WP_HTML_Processor_State::INSERTION_MODE_IN_TEMPLATE;
+				}
 
-		// When moving forwards, reparse the document until reaching the same location as the original bookmark.
-		if ( $bookmark_starts_at === $this->bookmarks[ $this->state->current_token->bookmark_name ]->start ) {
-			return true;
+				$this->reset_insertion_mode_appropriately();
```
This logic is inspired by https://github.com/WordPress/wordpress-develop/pull/7348 where a fragment parser is created on arbitrary context nodes and must be initialized properly.

This is very similar, the state must be reset to the appropriate initial state for the context.

This follows the fragment parsing algorithm: https://html.spec.whatwg.org/multipage/parsing.html#html-fragment-parsing-algorithm

sirreal committed:
Add and improve HTML processor bookmark tests

sirreal committed:
Add test seeking from SVG namespace

sirreal committed:
Rework with tests for fragment and full processors

Use a dataprovider to get a factory function for processors

sirreal committed:
Restore private stack properties to private

sirreal committed:
Use a temporary bookmark to avoid modifying tag processor internal state

sirreal commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -5343,32 +5342,65 @@ public function seek( $bookmark_name ): bool {
+			}
 			if ( $bookmark_starts_at === $this->bookmarks[ $this->state->current_token->bookmark_name ]->start ) {
-				while ( isset( $this->current_element ) && WP_HTML_Stack_Event::POP === $this->current_element->operation ) {
-					$this->current_element = array_shift( $this->element_queue );
```
I'd like special scrutiny here. I'm not sure why this section existed, it may have been to avoid stopping on virtual tokens.

The changes handle the directly by skipping virtual tokens and proceeding past them to reach the desired bookmark.

CC @dmsnell

dmsnell replied:
yes I suppose that it's true it's there to skip virtual nodes. I don't think that I thought of it that way when we put it together the first time.

there may be multiple nodes which start where the bookmark starts. I guess that these would necessarily be virtual closing tags. and the purpose is to slide forward until we reach the actual opening tag for the bookmark.

sirreal commented:
This is ready for review @gziolo @westonruter @dmsnell.

dmsnell commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -5324,8 +5324,7 @@ public function seek( $bookmark_name ): bool {
+			 * When moving backward, stateful stacks should be cleared.
 			 */
 			foreach ( $this->state->stack_of_open_elements->walk_up() as $item ) {
 				if ( 'context-node' === $item->bookmark_name ) {
```
doesn't this also need to go? I can’t remember, but I thought that we stopped pushing the context node onto the stack of open elements, and likely forgot to remove this when we did. same deal a few lines below.

sirreal replied:
I think you're right, good spot! I also think there was a bug due to this where it was supposed to break at the context node, but didn't, so ended up popping even the root HTML node that should always be present in the fragment parser!

I'll push a fix.

sirreal replied:
I've removed a few special cases that handled `context-node` on the open element stack, thanks for spotting that 👍

sirreal committed:
Break at root-node when popping elements on seek

The context node is not pushed to the stack of open elements,
this code was either doing nothing or incorrect.

sirreal committed:
Remove context-node cases from open elements

The context node is not pushed to the stack of open elements, and
therefore does not need special handling.

sirreal committed:
Merge branch 'trunk' into html-api/fix-63390-seek

sirreal commented:
@mi5t4n you were active on https://core.trac.wordpress.org/ticket/62290, if you could provide testing for this change it would help move things along 🙏

sirreal committed:
Fix root-node pop regression in fragment parser seek

sirreal committed:
Add test for root-node regression

sirreal commented:
Testing in https://github.com/WordPress/wordpress-develop/pull/7742 discovered an issue present on `trunk` where the fragment parser root-node may not be present on the stack of open elements in a fragment parser.

I've pushed a test and a fix for that to this PR.

sirreal committed:
Merge branch 'trunk' into html-api/fix-63390-seek

ockham commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -5328,52 +5328,92 @@ public function seek( $bookmark_name ): bool {
+			 * This must be done after clearing the stack because those stacks generate events that
+			 * would appear on a subsequent call to `next_token()`.
+			 */
+			$this->state->frameset_ok                       = true;
+			$this->state->stack_of_template_insertion_modes = array();
+			$this->state->head_element                      = null;
+			$this->state->form_element                      = null;
+			$this->state->current_token                     = null;
```
Would it make sense to add a `reset()` method to `WP_HTML_Processor_State`?

sirreal replied:
I considered that but ultimately decided against it. This isn't really a full reset, and the reset behaves a bit differently between the fragment/full processors.

To simplify this, I'd probably create a new state instance and copy over the important properties. That may be worth exploring but that felt like a larger refactor.

ockham reviewed (approved):
This is looking good to me! Thank you for the helpful inline comments and unit test coverage 👍 

I'm going to commit this later today (unless someone else beats me to it).

ockham commented:
Committed to Core `trunk` in https://core.trac.wordpress.org/changeset/59391.
